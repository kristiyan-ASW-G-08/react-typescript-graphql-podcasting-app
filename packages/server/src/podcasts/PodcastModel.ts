import mongoose, { Schema } from 'mongoose';
import PodcastType from '@customTypes/PodcastType';
import uniqueValidator from 'mongoose-unique-validator';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';

const PodcastSchema = new Schema<PodcastType>({
  title: { type: String, maxlength: 50, minlength: 1, unique: true },
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  website: { type: String, required: true },
  cover: { type: String, required: true },
});

// @ts-ignore
PodcastSchema.plugin(uniqueValidator);
// @ts-ignore
PodcastSchema.post('save', duplicationErrorHandler);
// @ts-ignore
PodcastSchema.post('update', duplicationErrorHandler);

PodcastSchema.index({ title: 'text' });

export default mongoose.model<PodcastType>('Podcast', PodcastSchema);
