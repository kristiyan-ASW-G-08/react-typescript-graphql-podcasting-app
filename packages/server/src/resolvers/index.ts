import userMutations from 'src/users/usersMutations';
import podcastMutations from '@src/podcasts/podcastsMutations';
import podcastQueries from '@src/podcasts/podcastQueries';
import episodeMutations from '@src/episodes/episodeMutations';

const resolvers = {
  Query: {
    ...podcastQueries,
  },
  Mutation: { ...userMutations, ...podcastMutations, ...episodeMutations },
};

export default resolvers;
