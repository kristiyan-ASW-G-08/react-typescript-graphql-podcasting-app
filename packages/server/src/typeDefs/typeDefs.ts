import { gql } from 'apollo-server';

const typeDefs = gql`
  scalar Upload

  type User {
    username: String!
    email: String!
    _id: ID!
  }

  type Query {
    getUser: User!
  }

  type UserLoginPayload {
    username: String!
    email: String!
    userId: String!
    token: String!
  }
  type Podcast {
    title: String!
    website: String!
    cover: String!
    user: User!
    _id: ID!
  }

  type Episode {
    title: String!
    description: String!
    audioFile: String!
    date: String!
    user: User!
    podcast: Podcast!
    _id: ID!
  }
  type Episodes {
    episodes: [Episode]!
    count: Int!
  }
  type Podcasts {
    podcasts: [Podcast]!
    count: Int!
  }
  type Query {
    getPodcasts(query: String, page: Int!, limit: Int!): Podcasts!
    getPodcastsByTitle(title: String!): Podcasts!
    getPodcast(podcastId: ID!): Podcast!
    getEpisode(episodeId: ID!): Episode!
    getEpisodesByPodcast(podcastId: ID!): Episodes!
  }
  type Mutation {
    createUserMutation(
      username: String!
      email: String!
      password: String!
      confirmPassword: String!
    ): UserLoginPayload!

    loginMutation(email: String!, password: String!): UserLoginPayload!

    createPodcastMutation(
      title: String!
      website: String!
      cover: Upload!
    ): Podcast!

    updatePodcastMutation(
      title: String!
      website: String!
      cover: Upload
      _id: ID!
    ): Podcast!

    createEpisodeMutation(
      title: String!
      description: String!
      audioFile: Upload!
      podcast: ID!
    ): Episode!
  }
`;

export default typeDefs;
