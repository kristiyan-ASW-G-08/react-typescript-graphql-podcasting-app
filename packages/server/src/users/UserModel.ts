import mongoose, { Schema } from 'mongoose';
import UserType from '@src/types/UserType';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: { type: String, minlength: 1, maxlength: 50, unique: true },
  email: { type: String, unique: true },
  password: { type: String, minlength: 8 },
});

UserSchema.plugin(uniqueValidator);
// @ts-ignore
UserSchema.post('save', duplicationErrorHandler);
// @ts-ignore
UserSchema.post('update', duplicationErrorHandler);

export default mongoose.model<UserType>('User', UserSchema);
