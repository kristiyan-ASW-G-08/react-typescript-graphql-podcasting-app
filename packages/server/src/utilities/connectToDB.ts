import mongoose from 'mongoose';
import logger from '@customUtilities/logger';

const connectToDB = async (mongoURI: string): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  } catch (err) {
    logger.error(`MongoDB Connection Error: ${err}`);
  }
};
export default connectToDB;
