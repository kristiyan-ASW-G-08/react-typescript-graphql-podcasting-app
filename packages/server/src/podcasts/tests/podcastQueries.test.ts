import mongoose from 'mongoose';
import { ApolloServer } from 'apollo-server-express';
import { sign } from 'jsonwebtoken';
import connectToDB from '@customUtilities/connectToDB';
import UserModel from '@src/users/UserModel';
import PodcastModel from '@src/podcasts/PodcastModel';
// @ts-ignore
import typeDefs from '@typeDefs/typeDefs';
import { Readable } from 'stream';
// @ts-ignore
import { Upload } from 'graphql-upload/public';
import resolvers from '../../resolvers/index';

describe('Resolvers', () => {
  const { MONGO_USER, MONGO_PASSWORD, MONGO_DATABASE } = process.env;
  const mongoURI = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@cluster0.ol9wi.mongodb.net/${MONGO_DATABASE}?retryWrites=true`;
  beforeAll(async () => {
    await mongoose.disconnect();
    await connectToDB(mongoURI);
    await UserModel.deleteMany({}).exec();
  });

  afterEach(async () => {
    await UserModel.deleteMany({}).exec();
  });
  afterAll(async () => {
    await mongoose.disconnect();
  });

  describe('Queries', () => {
    afterEach(async () => {
      await PodcastModel.deleteMany({}).exec();
    });

    describe('getPodcast', () => {
      it('should return a podcast successfully', async () => {
        expect.assertions(2);
        const podcast = await new PodcastModel({
          title: 'someTitle',
          website: 'someWebsite',
          user: mongoose.Types.ObjectId(),
          cover: 'someCover',
        }).save();
        const podcastId = podcast._id.toString();
        // @ts-ignore
        const query = `
          query getPodcast( $podcastId:ID!) {
            getPodcast(podcastId: $podcastId) {
              _id
            }
          }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            podcastId,
          },
        });
        expect(result.errors).toBeUndefined();
        expect(result.data?.getPodcast._id).toBeTruthy();
      });

      it('should return a 404 not found error', async () => {
        expect.assertions(1);
        const podcastId = mongoose.Types.ObjectId().toString();
        // @ts-ignore
        const query = `
          query getPodcast( $podcastId:ID!) {
            getPodcast(podcastId: $podcastId) {
              _id
            }
          }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            podcastId,
          },
        });
        expect(result.errors).toMatchSnapshot();
      });
    });

    describe('getPodcasts', () => {
      it('should return  podcasts successfully', async () => {
        const podcast = {
          website: 'someWebsite',
          cover: 'someCover',
        };
        const podcasts = [
          { ...podcast, title: 'someTitle1', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle2', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle3', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle4', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle5', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle6', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle7', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle8', user: mongoose.Types.ObjectId() },
        ];
        await PodcastModel.insertMany(podcasts);
        expect.assertions(2);
        // @ts-ignore
        const query = `
        query getPodcasts($page: Int!, $limit: Int!) {
          getPodcasts(page: $page, limit: $limit) {
            podcasts {
              title
              website
              cover
              _id
            }
          }
        }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            page: 1,
            limit: 5,
          },
        });
        expect(result.errors).toBeUndefined();
        expect(result.data?.getPodcasts.podcasts).toHaveLength(5);
      });
    });

    describe('getPodcastsByTitle', () => {
      it('should return  podcasts successfully', async () => {
        const podcast = {
          website: 'someWebsite',
          cover: 'someCover',
        };
        const podcasts = [
          {
            ...podcast,
            title: 'correctTitle1',
            user: mongoose.Types.ObjectId(),
          },
          {
            ...podcast,
            title: 'correctTitle2',
            user: mongoose.Types.ObjectId(),
          },
          {
            ...podcast,
            title: 'correctTitle3',
            user: mongoose.Types.ObjectId(),
          },
          {
            ...podcast,
            title: 'correctTitle4',
            user: mongoose.Types.ObjectId(),
          },
          { ...podcast, title: 'someTitle5', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle6', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle7', user: mongoose.Types.ObjectId() },
          { ...podcast, title: 'someTitle8', user: mongoose.Types.ObjectId() },
        ];
        await PodcastModel.insertMany(podcasts);
        expect.assertions(2);
        // @ts-ignore
        const query = `
        query getPodcastsByTitle($title:String!) {
          getPodcastsByTitle(title: $title) {
            podcasts {
              title
            }
          }
        }
        `;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
          variables: {
            title: 'correctTitle',
          },
        });
        expect(result.errors).toBeUndefined();
        expect(result.data?.getPodcastsByTitle.podcasts).toHaveLength(4);
      });
    });
  });
});
