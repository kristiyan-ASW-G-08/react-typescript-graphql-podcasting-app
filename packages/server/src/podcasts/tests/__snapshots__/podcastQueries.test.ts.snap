// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers Queries getPodcast should return a 404 not found error 1`] = `
Array [
  [GraphQLError: Podcast Was Not Found],
]
`;
